### ISSUE TITLE
Hyperparameter Setting

### ISSUE THREAD

**Issue by luyy9apples** [ema-marconato/reasoning-shortcuts#3]
Hi, could you provide the best hyperparameter settings? The current exp_best_args.py appears incomplete.
Additionally, I cannot reproduce the MNIST-EvenOdd results for the LTN model—training collapses to mapping only one digit. Could you advise on the cause?

**Comment by ema-marconato** [comment]
Hi,
thank you for the interest in the paper and on rerunning the experiments.

> could you provide the best hyperparameter settings? The current
exp_best_args.py appears incomplete.

Can you point to what hyperparameters are missing or what is not working?
Unfortunately I may have missed reporting some of them, but if you can
point to it, I will try fix the error.

> Additionally, I cannot reproduce the MNIST-EvenOdd results for the LTN
model—training collapses to mapping only one digit. Could you advise on the
cause?

Training with LTN is sometimes difficult. The issue is that it utilizes a
hyperparameter, model.p, which regulates the "strength" of the fuzzy
implementation of logical connectives. How to regulate it is mostly by
heuristics. The collapse you observe may be because the loss cannot
appropriately steer the concepts outside local minima. You can try changing
here:

https://github.com/ema-marconato/reasoning-shortcuts/blob/91b3e4231bb83e370eca3863e4a1125b877b20c9/XOR_MNIST/utils/ltn_loss.py#L28

the update is based on the epochs of training. Try also different seeds and
network intializations, that can help too!
Unfortunately, there is no general recipe for making LTN work. You can also
check the original implementation and see similar heuristics:

https://github.com/tommasocarraro/LTNtorch/blob/main/examples/4-semi-supervised_pattern_recognition.ipynb

Hope it helps!
Best,

Emanuele Marconato

P.S. I won't have access to github till next week. That means I cannot
solve your issue by then, but I can try help debugging the problem.




Il giorno mer 18 giu 2025 alle ore 05:39 luyy9apples <
***@***.***> ha scritto:

**Comment by luyy9apples** [comment]
Thanks for the detailed reply! I couldn't find the optimal learning rates and n_epoch for each model when not applying mitigation strategies.

**Comment by ema-marconato** [comment]
I see, thank you for pointing to that.
My bold guess is that the hyperparameters from addmnist should work.
You can check them here:

https://github.com/samuelebortolotti/bears/blob/99c51cd0c2520e0e317a43548616c8d54a0ae8a4/XOR_MNIST/exp_best_args.py#L4

Let me know if you find problems.

Emanuele

Il giorno gio 19 giu 2025 alle ore 07:18 luyy9apples <
***@***.***> ha scritto:

**Comment by luyy9apples** [comment]
I'm so sorry for the late reply. I've been trying to train LTN recently, but I haven't found a stable way yet. Thanks a lot for your help anyway!

Additionally, I've also run into some confusion while trying to reproduce the BDD-OIA experiments. After running the run_bdd.sh script to train the model, the calculated F1all(Y) and mF1(Y) values are 0.63 and 0.57, respectively, which are both a bit lower than the values reported in the paper. Could you please clarify how the F1 scores in the paper were calculated? Is it possible that the script's parameters were configured incorrectly?

**Comment by ema-marconato** [comment]
I guess for LTN you should try the implementation of rsbench (
https://github.com/unitn-sml/rsbench) and try to see if you can find a
solution training there.
If the issue persists, I suggest you to isolate the code on LTN that you
need and contacting Tommaso Carraro, the creator of LTNtorch.

For BDD-OIA, have you measured mean F1 according to the procedure here
https://arxiv.org/abs/2202.01459 ?
I am not sure where the problem is; it should be enough to have 5 epochs to
converge, but pay attention if some overfitting is in place.


Il giorno dom 31 ago 2025 alle ore 08:16 luyy9apples <
***@***.***> ha scritto:
